File: cpp/test/benchmark.cc
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.#include "benchmark.h"

// Some very basic throughput benchmarks.

#include "benchmark.h"

#include <functional>
#include <optional>

#include "factories.h"
#include "format.h"
#include "generated/binary/protocols.h"
#include "generated/hdf5/protocols.h"
#include "generated/ndjson/protocols.h"

using namespace test_model;

using namespace yardl::testing;

namespace {

Result BenchmarkFloat256x256(Format format) {
  auto scale = [format]() -> double {
    switch (format) {
      case Format::kNDJson:
        return 0.01;
      default:
        return 1;
    }
  }();

  yardl::FixedNDArray<float, 256, 256> a;
  int i = 0;
  for (auto& x : a) {
    x = static_cast<float>(++i) - std::numeric_limits<float>::epsilon();
  }

  size_t const repetitions = ScaleRepetitions(10000, scale);
  size_t const total_size = sizeof(a) * repetitions;

  return TimeScenario(
      total_size,
      [&]() {
        auto writer = CreateWriter<BenchmarkFloat256x256WriterBase>(format, kOutputFileName);
        for (size_t i = 0; i < repetitions; ++i) {
          writer->WriteFloat256x256(a);
        }
        writer->EndFloat256x256();
      },
      [&]() {
        auto reader = CreateReader<BenchmarkFloat256x256ReaderBase>(format, kOutputFileName);
        size_t read_repetitions = 0;
        while (reader->ReadFloat256x256(a)) {
          read_repetitions++;
        }
        AssertRepetitionsSame(repetitions, read_repetitions);
      });
}

Result BenchmarkFloatVlen(Format format) {
  auto scale = [format]() -> double {
    switch (format) {
      case Format::kHdf5:
        return 0.5;
      case Format::kNDJson:
        return 0.01;
      default:
        return 1;
    }
  }();

  yardl::NDArray<float, 2> a({256, 256});
  int i = 0;
  for (auto& x : a) {
    x = static_cast<float>(++i) - std::numeric_limits<float>::epsilon();
  }

  size_t const repetitions = ScaleRepetitions(10000, scale);
  size_t const total_size = sizeof(float) * a.size() * repetitions;

  return TimeScenario(
      total_size,
      [&]() {
        auto writer = CreateWriter<BenchmarkFloatVlenWriterBase>(format, kOutputFileName);
        for (size_t i = 0; i < repetitions; ++i) {
          writer->WriteFloatArray(a);
        }
        writer->EndFloatArray();
      },
      [&]() {
        auto reader = CreateReader<BenchmarkFloatVlenReaderBase>(format, kOutputFileName);
        size_t read_repetitions = 0;
        while (reader->ReadFloatArray(a)) {
          read_repetitions++;
        }
        AssertRepetitionsSame(repetitions, read_repetitions);
      });
}

Result BenchmarkSmallInt256x256(Format format) {
  auto scale = [format]() -> double {
    switch (format) {
      case Format::kNDJson:
        return 0.01;
      default:
        return 1;
    }
  }();

  yardl::FixedNDArray<int, 256, 256> a;
  for (auto& x : a) {
    x = 37;
  }

  size_t const repetitions = ScaleRepetitions(6000, scale);
  size_t const total_size = sizeof(a) * repetitions;

  return TimeScenario(
      total_size,
      [&]() {
        auto writer = CreateWriter<BenchmarkInt256x256WriterBase>(format, kOutputFileName);
        for (size_t i = 0; i < repetitions; ++i) {
          writer->WriteInt256x256(a);
        }
        writer->EndInt256x256();
      },
      [&]() {
        auto reader = CreateReader<BenchmarkInt256x256ReaderBase>(format, kOutputFileName);
        size_t read_repetitions = 0;
        while (reader->ReadInt256x256(a)) {
          read_repetitions++;
        }
        AssertRepetitionsSame(repetitions, read_repetitions);
      });
}

Result BenchmarkSmallRecord(Format format) {
  auto scale = [format]() -> double {
    switch (format) {
      case Format::kHdf5:
        return 0.005;
      case Format::kNDJson:
        return 0.01;
      default:
        return 1;
    }
  }();

  SmallBenchmarkRecord record{73278383.23123213, 78323.2820379, -2938923.29882};

  size_t const repetitions = ScaleRepetitions(50000000, scale);
  static_assert(sizeof(record) == 16);
  size_t const total_size = sizeof(record) * repetitions;

  return TimeScenario(
      total_size,
      [&]() {
        auto writer = CreateWriter<BenchmarkSmallRecordWriterBase>(format, kOutputFileName);
        for (size_t i = 0; i < repetitions; ++i) {
          writer->WriteSmallRecord(record);
        }
        writer->EndSmallRecord();
      },
      [&]() {
        auto reader = CreateReader<BenchmarkSmallRecordReaderBase>(format, kOutputFileName);
        size_t read_repetitions = 0;
        while (reader->ReadSmallRecord(record)) {
          read_repetitions++;
        }
        AssertRepetitionsSame(repetitions, read_repetitions);
      });
}

Result BenchmarkSmallRecordBatched(Format format) {
  auto scale = [format]() -> double {
    switch (format) {
      case Format::kNDJson:
        return 0.005;
      default:
        return 1;
    }
  }();

  SmallBenchmarkRecord const record{73278383.23123213, 78323.2820379, -2938923.29882};
  std::vector<SmallBenchmarkRecord> batch(8192, record);

  size_t const repetitions = ScaleRepetitions(20000, scale);
  size_t const total_size = batch.size() * sizeof(record) * repetitions;

  return TimeScenario(
      total_size,
      [&]() {
        auto writer = CreateWriter<BenchmarkSmallRecordWriterBase>(format, kOutputFileName);
        for (size_t i = 0; i < repetitions; ++i) {
          writer->WriteSmallRecord(batch);
        }
        writer->EndSmallRecord();
      },
      [&]() {
        auto reader = CreateReader<BenchmarkSmallRecordReaderBase>(format, kOutputFileName);
        size_t read_repetitions = 0;
        while (reader->ReadSmallRecord(batch)) {
          read_repetitions++;
        }
        AssertRepetitionsSame(repetitions, read_repetitions);
      });
}

Result BenchmarkSmallOptionalsBatched(Format format) {
  auto scale = [format]() -> double {
    switch (format) {
      case Format::kNDJson:
        return 0.01;
      default:
        return 1;
    }
  }();

  SimpleEncodingCounters const record{26723, 92738, 7899};
  std::vector<SimpleEncodingCounters> batch(8192, record);

  size_t repetitions = ScaleRepetitions(5000, scale);
  size_t total_size = batch.size() * sizeof(record) * repetitions;

  return TimeScenario(
      total_size,
      [&]() {
        auto writer = CreateWriter<BenchmarkSmallRecordWithOptionalsWriterBase>(format, kOutputFileName);
        for (size_t i = 0; i < repetitions; ++i) {
          writer->WriteSmallRecord(batch);
        }
        writer->EndSmallRecord();
      },
      [&]() {
        auto reader = CreateReader<BenchmarkSmallRecordWithOptionalsReaderBase>(format, kOutputFileName);
        size_t read_repetitions = 0;
        while (reader->ReadSmallRecord(batch)) {
          read_repetitions++;
        }
        AssertRepetitionsSame(repetitions, read_repetitions);
      });
}

Result BenchmarkSimpleMrd(Format format) {
  auto scale = [format]() -> double {
    switch (format) {
      case Format::kHdf5:
        return 0.5;
      case Format::kNDJson:
        return 0.01;
      default:
        return 1;
    }
  }();

  SimpleAcquisition acq;
  acq.data.resize({32, 256});
  acq.trajectory = yardl::NDArray<float, 2>({32, 2});
  std::variant<SimpleAcquisition, Image<float>> value = acq;

  size_t const repetitions = ScaleRepetitions(30000, scale);
  size_t single_size = sizeof(value) + acq.data.size() * sizeof(std::complex<float>) + acq.trajectory.size() * sizeof(float);
  if (single_size != 66032) {
    throw std::runtime_error("Unexpected size: " + std::to_string(single_size));
  }
  size_t const total_size = single_size * repetitions;

  return TimeScenario(
      total_size,
      [&]() {
        auto writer = CreateWriter<BenchmarkSimpleMrdWriterBase>(format, kOutputFileName);
        for (size_t i = 0; i < repetitions; ++i) {
          writer->WriteData(value);
        }
        writer->EndData();
      },
      [&]() {
        auto reader = CreateReader<BenchmarkSimpleMrdReaderBase>(format, kOutputFileName);
        size_t read_repetitions = 0;
        while (reader->ReadData(value)) {
          read_repetitions++;
        }
        AssertRepetitionsSame(repetitions, read_repetitions);
      });
}

std::map<std::string, std::function<Result(Format)>>
    function_map = {
        {"float256x256", BenchmarkFloat256x256},
        {"smallint256x256", BenchmarkSmallInt256x256},
        {"floatvlen", BenchmarkFloatVlen},
        {"smallrecord", BenchmarkSmallRecord},
        {"smallrecordbatched", BenchmarkSmallRecordBatched},
        {"smalloptionalsbatched", BenchmarkSmallOptionalsBatched},
        {"simplemrd", BenchmarkSimpleMrd},
};

std::optional<std::function<Result(Format)>> GetScenarioFunction(std::string scenario) {
  std::transform(scenario.begin(), scenario.end(), scenario.begin(), [](unsigned char c) { return std::tolower(c); });

  auto func_pair = function_map.find(scenario);
  if (func_pair == function_map.end()) {
    return std::nullopt;
  }
  return func_pair->second;
}

}  // namespace

int main(int argc, char* argv[]) {
  if (argc != 3) {
    std::cerr << "Incorrect number of arguments. Usage: banchmark <scenario> <hdf5 | binary | ndjson>" << std::endl;
    return 1;
  }

  std::string scenario = argv[1];
  Format format = ParseFormat(std::string(argv[2]));
  auto func = GetScenarioFunction(scenario);
  if (!func) {
    std::cerr << "Unknown scenario: " << scenario << std::endl;
    return 0;
  }

  Result result = (*func)(format);

  nlohmann::ordered_json j = result;
  std::cout << j << std::endl;
}

File: matlab/test/benchmark.m
% Copyright (c) Microsoft Corporation.
% Licensed under the MIT License.

function benchmark(varargin)
    if nargin ~= 2
        error("Usage: benchmark <scenario> <format>");
    end

    scenario = parse_scenario(varargin{1});
    format = parse_format(varargin{2});

    % Matlab support currently implemented for binary format only
    if format ~= "binary"
        return
    end

    addpath("../generated/");
    fprintf("%s\n", scenario(format));
end

function output_filename = get_output_filename()
    output_filename = "/tmp/benchmark_data.dat";
end

function format = parse_format(format)
    switch format
        case "binary"
            format = "binary";
        case "ndjson"
            format = "ndjson";
        case "hdf5"
            format = "hdf5";
        otherwise
            error("Invalid format");
    end
end

function scenario = parse_scenario(scenario)
    switch scenario
        case "float256x256"
            scenario = @benchmark_float256x256;
        case "floatvlen"
            scenario = @benchmark_float_vlen;
        case "smallint256x256"
            scenario = @benchmark_small_int_256x256;
        case "smallrecord"
            scenario = @benchmark_small_record;
        case "smallrecordbatched"
            scenario = @benchmark_small_record_batched;
        case "smalloptionalsbatched"
            scenario = @benchmark_small_optionals_batched;
        case "simplemrd"
            scenario = @benchmark_simple_mrd;
        otherwise
            error("Invalid scenario");
    end
end

function [writer, reader] = create_writer_reader(format, protocol)
    format_name = "test_model." + format + "." + protocol;
    writer_name = format_name + "Writer";
    reader_name = format_name + "Reader";
    writer = str2func(writer_name);
    reader = str2func(reader_name);
end

function w = create_validating_writer(testCase, format, protocol)
    writer_name = "test_model." + format + "." + protocol + "Writer";
    reader_name = "test_model." + format + "." + protocol + "Reader";
    test_writer_name = "test_model.testing.Test" + protocol + "Writer";

    create_writer = str2func(writer_name);
    create_reader = str2func(reader_name);
    create_test_writer = str2func(test_writer_name);

    w = create_test_writer(testCase, format, @(f) create_writer(f), @(f) create_reader(f));
end

function reps = scale_repetitions(repetitions, scale)
    reps = repetitions * scale;
end

function res = time_scenario(total_bytes_size, write_func, read_func)
    delete(get_output_filename());

    total_size_mi_byte = total_bytes_size / 1024 / 1024;

    write_start = tic;
    write_func();
    write_elapsed = toc(write_start);
    write_mibps = total_size_mi_byte / write_elapsed;

    read_start = tic;
    read_func();
    read_elapsed = toc(read_start);
    read_mibps = total_size_mi_byte / read_elapsed;

    s.write_mi_bytes_per_second = write_mibps;
    s.read_mi_bytes_per_second = read_mibps;
    s.roundtrip_duration_seconds = write_elapsed + read_elapsed;
    res = jsonencode(s);
end

function res = benchmark_float256x256(format)
    scale = 1;

    a = zeros(256, 256, 'single');
    for i = 1:numel(a)
        a(i) = i - eps;
    end

    repetitions = scale_repetitions(10000, scale);
    total_bytes = 4 * numel(a) * repetitions;

    [create_writer, create_reader] = create_writer_reader(format, 'BenchmarkFloat256x256');

    function write()
        w = create_writer(get_output_filename());
        for r = 1:repetitions
            w.write_float256x256({a});
        end
        w.close();
    end

    function read()
        r = create_reader(get_output_filename());
        count = 0;
        while r.has_float256x256()
            b = r.read_float256x256();
            count = count + 1;
        end
        assert(count == repetitions);
        r.close();
    end

    res = time_scenario(total_bytes, @write, @read);
end

function res = benchmark_float_vlen(format)
    scale = 1;

    a = zeros(256, 256, 'single');
    for i = 1:numel(a)
        a(i) = i - eps;
    end

    repetitions = scale_repetitions(10000, scale);
    total_bytes = 4 * numel(a) * repetitions;

    [create_writer, create_reader] = create_writer_reader(format, 'BenchmarkFloatVlen');

    function write()
        w = create_writer(get_output_filename());
        for r = 1:repetitions
            w.write_float_array({a});
        end
        w.close();
    end

    function read()
        r = create_reader(get_output_filename());
        count = 0;
        while r.has_float_array()
            b = r.read_float_array();
            count = count + 1;
        end
        assert(count == repetitions);
        r.close();
    end

    res = time_scenario(total_bytes, @write, @read);
end

function res = benchmark_small_int_256x256(format)
    scale = 1;

    a = zeros(256, 256, 'int32');
    for i = 1:numel(a)
        a(i) = int32(37);
    end

    repetitions = scale_repetitions(10, scale);
    total_bytes = 4 * numel(a) * repetitions;

    [create_writer, create_reader] = create_writer_reader(format, 'BenchmarkInt256x256');

    function write()
        w = create_writer(get_output_filename());
        for r = 1:repetitions
            w.write_int256x256({a});
        end
        w.close();
    end

    function read()
        r = create_reader(get_output_filename());
        count = 0;
        while r.has_int256x256()
            b = r.read_int256x256();
            count = count + 1;
        end
        assert(count == repetitions);
        r.close();
    end

    res = time_scenario(total_bytes, @write, @read);
end

function res = benchmark_small_record(format)
    scale = 1;

    record = test_model.SmallBenchmarkRecord(double(73278383.23123213), single(78323.2820379), single(-2938923.29882));

    repetitions = scale_repetitions(50000, scale);
    total_bytes = 16 * repetitions;

    [create_writer, create_reader] = create_writer_reader(format, 'BenchmarkSmallRecord');

    function write()
        w = create_writer(get_output_filename());
        for r = 1:repetitions
            w.write_small_record({record});
        end
        w.close();
    end

    function read()
        r = create_reader(get_output_filename());
        count = 0;
        while r.has_small_record()
            b = r.read_small_record();
            count = count + 1;
        end
        assert(count == repetitions);
        r.close();
    end

    res = time_scenario(total_bytes, @write, @read);
end

function res = benchmark_small_record_batched(format)
    % Batching has not yet been implemented for Matlab
    res = "";
end

function res = benchmark_small_optionals_batched(format)
    % Batching has not yet been implemented for Matlab
    res = "";
end

function res = benchmark_simple_mrd(format)
    scale = 1;

    acq = test_model.SimpleAcquisition();
    acq.data = complex(zeros(256, 32, 'single'));
    acq.trajectory = zeros(2, 32, 'single');

    value = test_model.AcquisitionOrImage.Acquisition(acq);

    repetitions = scale_repetitions(2500, scale);
    total_bytes = 66032 * repetitions;

    [create_writer, create_reader] = create_writer_reader(format, 'BenchmarkSimpleMrd');

    function write()
        w = create_writer(get_output_filename());
        for r = 1:repetitions
            w.write_data({value});
        end
        w.close();
    end

    function read()
        r = create_reader(get_output_filename());
        count = 0;
        while r.has_data()
            b = r.read_data();
            count = count + 1;
        end
        assert(count == repetitions);
        r.close();
    end

    res = time_scenario(total_bytes, @write, @read);
end
